mlir/lib/Catalyst/Transforms/DetectQNodes.cpp

```diff
@@ -0,0 +1,1272 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Analysis/Liveness.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+
+#include "mlir/Dialect/LLVMIR/FunctionCallUtils.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Catalyst/Transforms/Passes.h"
+#include "Catalyst/Transforms/Patterns.h"
+
+using namespace mlir;
+
+namespace {
+
+// Constants
+
+static constexpr llvm::StringRef qnodeAttr = "qnode";
```

Comment:
nice! didn't know that you can use `llvm::StringRef` with `constexpr` when building with C++17. It's probably implemented as `std::string_view` instead of `std::string` :thinking: as you can use `constexpr` with `string_view` in C++17. `constexpr std::string` is a C++20 feature.